src/schema/schema_alter.c

```diff
@@ -63,7 +54,164 @@ err:	__wt_free(session, config);
+	 * closed after we're done with it and the next open will see the
+	 * updated metadata.
+	 */
+	filename = uri = session->dhandle->name;
+	if (!WT_PREFIX_SKIP(filename, "file:"))
+		return (__wt_unexpected_object_type(session, uri, "file:"));
+
+	WT_RET(__wt_meta_track_on(session));
 
+	WT_ERR(__alter_apply(session,
+	    uri, cfg, WT_CONFIG_BASE(session, file_meta)));
+
+err:	WT_TRET(__wt_meta_track_off(session, true, ret != 0));
 	return (ret);
 }
+
+/*
+ * __alter_tree --
+ *	Alter an index or colgroup reference.
+ */
+static int
+__alter_tree(WT_SESSION_IMPL *session, const char *name, const char *cfg[])
+{
+	WT_CONFIG_ITEM cval;
+	WT_DECL_ITEM(data_source);
+	WT_DECL_RET;
+	char *value;
+	bool is_colgroup;
+
+	value = NULL;
+
+	is_colgroup = WT_PREFIX_MATCH(name, "colgroup:");
+	if (!is_colgroup && !WT_PREFIX_MATCH(name, "index:"))
+		WT_ERR_MSG(session, EINVAL,
+		    "expected a 'colgroup:' or 'index:' source: '%s'", name);
+
+	/* Read the schema value. */
+	WT_ERR(__wt_metadata_search(session, name, &value));
+
+	/* Get the data source URI. */
+	if ((ret = __wt_config_getones(session, value, "source", &cval)) != 0)
+		WT_ERR_MSG(session, EINVAL,
+		    "index or column group has no data source: %s", value);
+
+	WT_ERR(__wt_scr_alloc(session, 0, &data_source));
+	WT_ERR(__wt_buf_fmt(session,
+	    data_source, "%.*s", (int)cval.len, cval.str));
+
+	/* Alter the data source */
+	WT_ERR(__wt_schema_alter(session, data_source->data, cfg));
+
+	/* Alter the index or colgroup */
+	if (is_colgroup)
+		WT_ERR(__alter_apply(session,
+		    name, cfg, WT_CONFIG_BASE(session, colgroup_meta)));
+	else
+		WT_ERR(__alter_apply(session,
+		    name, cfg, WT_CONFIG_BASE(session, index_meta)));
+
+err:	__wt_scr_free(session, &data_source);
+	__wt_free(session, value);
+	return (ret);
+}
+
+/*
+ * __alter_table --
+ *	WT_SESSION::alter for a table.
+ */
+static int
+__alter_table(
+    WT_SESSION_IMPL *session, const char *uri, const char *cfg[])
+{
+	WT_COLGROUP *colgroup;
+	WT_DECL_RET;
+	WT_INDEX *idx;
+	WT_TABLE *table;
+	u_int i;
+	const char *name;
+	bool tracked;
+
+	name = uri;
+	WT_PREFIX_SKIP_REQUIRED(session, name, "table:");
+
+	table = NULL;
+	tracked = false;
+
+	/*
+	 * Open the table so we can alter its column groups and indexes, keeping
+	 * the table locked exclusive across the alter.
+	 */
+	WT_ERR(__wt_schema_get_table_uri(session, uri, true,
+	    WT_DHANDLE_EXCLUSIVE, &table));
+
+	colgroup = NULL;
+	/* Alter the column groups. */
+	for (i = 0; i < WT_COLGROUPS(table); i++) {
+		if ((colgroup = table->cgroups[i]) == NULL)
+			continue;
+		WT_ERR(__alter_tree(session, colgroup->name, cfg));
+	}
+
+	/* Alter the indices. */
+	WT_ERR(__wt_schema_open_indices(session, table));
+	for (i = 0; i < table->nindices; i++) {
+		if ((idx = table->indices[i]) == NULL)
+			continue;
+		WT_ERR(__alter_tree(session, idx->name, cfg));
+	}
+
+	/* Alter the table */
+	WT_ERR(__alter_apply(session,
```

Comment:
Unlike rename_table, why alter is done before __wt_meta_track_handle_lock? Just curious.

Reply:
It benefits from releasing lock early. Rename is multistage, because it has to rename the underlying files/sources and then the metadata. Metadata rename is really search, remove-old and insert-new and can be done without metatracking under lock. Alter is simpler and is mostly just modifying configuration for the tables. Being just a single step, it needs to be done under metatracking.